Add support for using GSSNTLMSSP module for NTLM and channel bindings

simo5 opened the pull request:
GSSNTLMSSP is a GSSAPI module that implement the NTLMSSP protocol described in MS-NLMP
These patches allow to configure it's use, this in turns enables the use of Channel Bindings with NTLM's HTTP auth.

This builds only with configure so far (sorry not a cmake expert) and channel bindings are implemented only when OpenSSL  is used.

Seeking initial review and feedback

bagder commented:
Remember to run `make checksrc` locally;
~~~
./curl_ntlm_core.c:516:83: warning: Longer than 79 columns (LONGLINE)
 #if defined(USE_NTLM_V2) && !defined(USE_WINDOWS_SSPI) && !defined(USE_GSSNTLMSSP)
./curl_ntlm_core.c:567:83: warning: Longer than 79 columns (LONGLINE)
 #if defined(USE_NTLM_V2) && !defined(USE_WINDOWS_SSPI) && !defined(USE_GSSNTLMSSP)
./curl_ntlm_core.h:75:83: warning: Longer than 79 columns (LONGLINE)
 #if defined(USE_NTLM_V2) && !defined(USE_WINDOWS_SSPI) && !defined(USE_GSSNTLMSSP)
checksrc: 0 errors and 3 warnings

bagder commented:
We need docs/explanations. What it is this? How do we use it? How do users enable it and why?

Can we add test cases verifying it?

bagder commented on lib/vauth/ntlm_gssntlmssp.c:
```diff
@@ -257,10 +266,13 @@ CURLcode Curl_auth_create_ntlm_type3_message(struct Curl_easy *data,
   (void) userp;
+  gss_channel_bindings_t cb = GSS_C_NO_CHANNEL_BINDINGS;
+
+  if(ntlm->cb.application_data.value) cb = &ntlm->cb;
```
split on two lines

simo5 replied:
will do

bagder commented on lib/vauth/ntlm_gssntlmssp.c:
```diff
@@ -118,10 +119,18 @@ CURLcode Curl_auth_create_ntlm_type1_message(struct Curl_easy *data,
+                                         CURL_SSL_CB_TLS_SERVER_END_POINT,
+                                         &ntlm->cb.application_data.value,
+                                         &ntlm->cb.application_data.length);
+  if(ntlm->cb.application_data.value) cb = &ntlm->cb;
```
should be two lines

simo5 replied:
ok

simo5 commented:
> Remember to run `make checksrc` locally;
> 
> ```
> ./curl_ntlm_core.c:516:83: warning: Longer than 79 columns (LONGLINE)
>  #if defined(USE_NTLM_V2) && !defined(USE_WINDOWS_SSPI) && !defined(USE_GSSNTLMSSP)
> ./curl_ntlm_core.c:567:83: warning: Longer than 79 columns (LONGLINE)
>  #if defined(USE_NTLM_V2) && !defined(USE_WINDOWS_SSPI) && !defined(USE_GSSNTLMSSP)
> ./curl_ntlm_core.h:75:83: warning: Longer than 79 columns (LONGLINE)
>  #if defined(USE_NTLM_V2) && !defined(USE_WINDOWS_SSPI) && !defined(USE_GSSNTLMSSP)
> checksrc: 0 errors and 3 warnings
> ```

I thought I already fixed this in the last rebase, sorry I'll rcheck.

simo5 commented:
> We need docs/explanations. What it is this? How do we use it? How do users enable it and why?
> 
> Can we add test cases verifying it?

gssntlmssp is a gssapi module that allows things like proper SPNEGO negotiation (krb5+ntlm fallback for example) using GSSAPI.
It is installed in distributions that provide it as the "gssntlmssp" package.  Among other things it integrates with Winbind when compiled with that support so technically it can do transparent NTLM auth in AD joined Domain.

I can probably add a test case, if it is ok to install packages in any of the supported CI hosts. But I am not familiar with curl's CI setup so any pointer would be nice.
note that gssntlmssp can also be used with mod_auth_gssapi in apache to provide the server side NTLM auth for tests.

simo5 commented on lib/vtls/vtls.c:
```diff
@@ -1164,7 +1180,9 @@ static const struct Curl_ssl Curl_ssl_multi = {
-  NULL                               /* sha256sum */
+  NULL,                              /* sha256sum */
+  NULL,                              /* get_tls_unique */
+  NULL,                              /* get_tls_endpoint */
```
These are needed to extract the data needed from openssl for channel bindings.
There are two kind of TLS channel bindings that can be used: tls-uniqe and tls-server-end-point, microsoft chose to use the latter for HTTP

simo5 commented:
Pushed fixes for the code style nits.

No tests/docs yet

simo5 commented:
Test 1538 checks CURLE errors, by adding a new error I am making this test fail, however it is not clear to me how I can fix the test, any hints?

jay commented:
> Test 1538 checks CURLE errors, by adding a new error I am making this test fail, however it is not clear to me how I can fix the test, any hints?

You'll have to add the new error string to the test
e97: Internal SSL library error
e98: Unknown error

simo5 commented:
Ok I moved the printing in strerror, should have realized before there were no breaks there.
I also fixed the one 2-lines breakup I missed and fixed the test1538 data file.
That should take care of the code style and errors in the tests.

So what about adding a test in for this code, is there a test that already sets up an apache environment I can modify/expand? Or would this need a new setup?

simo5 commented:
Uhmm seem X509_get0_signature() was not available before openssl 1.0.2 ... I'll provide a patch for that shortly

simo5 commented:
@bagder so I have a problem here, I finally activated a travis test that uses gss-ntlmssp, however this reveals that test 2031 is bogus, as it tries to compare NTLM tokens directly.
BUT the NTLMSSP protocol is not as deterministic as the test would like, within NTLM implementations you can have various options and features turned on/off, and clearly gssntlmssp has a different feature set turned on by default compared to the default implementation in curl.
It's not clear to me how to fix it, is there a way to mark some output to be matched an other not to be matched ?

simo5 commented:
There is also this error in some CI runs:

```
vtls/openssl.c:4188:29: error: passing 'const X509_ALGOR **' (aka 'const struct

      X509_algor_st **') to parameter of type 'X509_ALGOR **' (aka 'struct

      X509_algor_st **') discards qualifiers in nested pointer types

      [-Werror,-Wincompatible-pointer-types-discards-qualifiers]

  X509_get0_signature(NULL, &sig_alg, server_cert);

                            ^~~~~~~~

/usr/include/openssl/x509.h:837:63: note: passing argument to parameter 'palg'

      here

void X509_get0_signature(ASN1_BIT_STRING **psig, X509_ALGOR **palg,

                                                              ^

1 error generated.
```

which is odd, because that same function is used in another part of the file with the same argument type and there is no error there ...
In openssl 1.1.1 the argument must be const and I get a compile error if I change it, so unclear how to deal with this across different builds...

bagder commented:
> test 2031 is bogus, as it tries to compare NTLM tokens directly.

I wouldn't call it "bogus" because of that. It is written in a way so that the data *is* comparable. Your patch obviously changes that, but then I suppose test 2031 needs to be adjusted to the new circumstances!

>  is there a way to mark some output to be matched an other not to be matched ?

It's a bit tricky. That test for example is explictitly not running for "!SSPI" so it is already disable if NTLM is done by the SSPI thing in Windows. Another way we do it in some tests is that we strip out certain parts of the output before it is compared (typically for example the `User-Agent` header) and in yet others we introduce %variables for parts we know will vary and we know the content.

Yet another approach we use is to make the test only run on debug versions of curl and then allow an environment variable override something internal to make the output more fixed and less random.

> There is also this error in some CI runs:

Curious, if I look at the headers for OpenSSL 1.1.1e on my dev box it shows this prototype:
~~~
void X509_get0_signature(const ASN1_BIT_STRING **psig,
                         const X509_ALGOR **palg, const X509 *x);

simo5 commented:
> > test 2031 is bogus, as it tries to compare NTLM tokens directly.
> 
> I wouldn't call it "bogus" because of that. It is written in a way so that the data _is_ comparable. Your patch obviously changes that, but then I suppose test 2031 needs to be adjusted to the new circumstances!

Didn't mean to pass judgement, just to highlight it's the test that is too picky and not a failure of the module.
 
> > is there a way to mark some output to be matched an other not to be matched ?
> 
> It's a bit tricky. That test for example is explictitly not running for "!SSPI" so it is already disable if NTLM is done by the SSPI thing in Windows. Another way we do it in some tests is that we strip out certain parts of the output before it is compared (typically for example the `User-Agent` header) and in yet others we introduce %variables for parts we know will vary and we know the content.

Would it make sense to add a !GSSNTLMSSP as well?
I am not sure I want to create another similar test for gssntlmssp though as I did change flags over time to make defaults more secure (as much as possible given the limitation of NTLMSSP).

> Yet another approach we use is to make the test only run on debug versions of curl and then allow an environment variable override something internal to make the output more fixed and less random.

If this is acceptable I can do this too, how is it done?

> > There is also this error in some CI runs:
> 
> Curious, if I look at the headers for OpenSSL 1.1.1e on my dev box it shows this prototype:
> 
> ```
> void X509_get0_signature(const ASN1_BIT_STRING **psig,
>                          const X509_ALGOR **palg, const X509 *x);
> ```

Exactly, in fact the code as it is now compiles just fine on most boxes, I yet have to figure out why some boxes raise that error, and why they do it only on the second invocation of X509_get0_signature() I added, but not the first that is already there...

simo5 commented:
I now also pushed a patch to make SSPI/Schannel use the same API just introduced for Channel Bindings instead of exporting internal data structures in various places.

I was wondering if I should split this into 2 separate PRs, onw for introducing gssntlmssp with commits 1 and 3 and a separate one to retool channel bindings ?

simo5 commented:
> Curious, if I look at the headers for OpenSSL 1.1.1e on my dev box it shows this prototype:
> 
> ```
> void X509_get0_signature(const ASN1_BIT_STRING **psig,
>                          const X509_ALGOR **palg, const X509 *x);
> ```

Found what "causes" it @bagder , this problem happens on configurations using xenial, which has openssl-1.0.2g and defines the function as:
```
void X509_get0_signature(ASN1_BIT_STRING **psig, X509_ALGOR **palg,
                         const X509 *x);
```

simo5 commented:
@bagder so build issues are now fixed, however my attempt to suppress NTLM tests when GSSNTLMSSP is configured, like for SSPI seem to have failed, and I can;t figure out why, as far as I can tell I am configuring this the same way it is done for SSPI.

bagder commented:
I still lack an explanation of what exactly this feature is for and how users enable it. Also, it needs a rebase and merge conflicts fixed!

simo5 commented:
@bagder this feature allows the use of NTLM through a GSSAPI module called gssntlmssp, this has several implications:
- gssntlmssp is integrated with winbindd, so you can transparently authenticate with --negotiate, without having to provide a password, instead winbindd will use the cached credentials obtained at login time
- a user-file can also be used (via env var) so ntlm authentication can be automated consistently across the system
- eventually: you can use negotiate and fallback from krb5 to ntlm transparently based on what credentials you have available in your system (for example if you leegd in on a system joined to a Windows domain via winbindd)
- you have a correct implementation of NTLM including all the securty levels and MIC protections, and potentially even channel bindings, and it is a fully tested implementation.

jay commented:
It seems like a lot of work was put into this and I don't want it to go under but we need someone on this project who is familiar with this protocol to review it.

simo5 commented:
Let me know if there is anything I can do to help here.

simo5 commented:
@jay any chance a reviewer will be assigned?
Otherwise I'd like to close and drop this work. I wish I had more time to dedicate but I don't.

jay commented:
> any chance a reviewer will be assigned?
> Otherwise I'd like to close and drop this work. I wish I had more time to dedicate but I don't.

@monnerat can you take this or who else knows about GSS and NTLM? @captain-caveman2k?

monnerat commented:
> can you take this or who else knows about GSS and NTLM?

No, I'm sorry: obviously the author knows a lot more about GSS that I do and it seems to require a strong infrastructure for real tests.

simo5 commented:
No problem in leaving the branch up.
And I may interact with this PR in my spare time, I will just not consider this as active work anymore.
Thanks.
